gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/CallStep.java

```diff
@@ -0,0 +1,303 @@
+        super(traversal);
+
+        this.isStart = isStart;
+        this.params = params == null ? new LinkedHashMap() : params;
+        this.mapTraversal = mapTraversal == null ? null : integrateChild(mapTraversal);
+        this.parameters = new Parameters();
+
+        // throws exception for unrecognized service
+        this.service = getServiceRegistry().get(service);
+
+        this.ctx = new ServiceCallContext(traversal, this);
+    }
+
+    @Override
+    protected Traverser.Admin<E> processNextStart() {
+        if (isStart && first) {
+            first = false;
+
+            /*
+             * Start of a traversal (no input). Obtain the one-time service iterator.
+             */
+
+            if (this.starts.hasNext()) {
+                throw new IllegalStateException("This service must be called without input: " + service.getName());
+            }
+
+            final Map params = getParams();
+            this.iterator = service.execute(this.ctx, params);
+        }
+
+        while (true) {
+            if (this.iterator.hasNext()) {
+
+                /*
+                 * Still draining the current iterator.
+                 */
+
+                final Object next = this.iterator.next();
+                if (next instanceof Traverser.Admin) {
+                    /*
+                     * Service is producing its own traversers (or is a pass-through on the input).
+                     */
+                    return (Traverser.Admin<E>) next;
+                } else if (this.head != null) {
+                    /*
+                     * Streaming service mapping input (head) to non-traverser output.
+                     */
+                    return this.head.split((E) next, this);
+                } else {
+                    /*
+                     * Barrier service producing non-traverser output, we have no head to split against. This loses
+                     * all path information.
+                     */
+                    return this.traversal.getTraverserGenerator().generate(next, (Step) this, 1l);
+                }
+            } else {
+
+                /*
+                 * Time to obtain another iterator from upstream input.
+                 */
+
+                closeIterator();
+                if (!this.starts.hasNext()) {
+                    // no more input
+                    throw FastNoSuchElementException.instance();
+                }
+                if (service.isBarrier()) {
+                    /*
+                     * Barrier service - gather all upstream input and call.
+                     */
+                    final TraverserSet<S> traverserSet = (TraverserSet<S>) this.traversal.getTraverserSetSupplier().get();
+                    this.starts.forEachRemaining(traverserSet::add);
+                    this.iterator = this.flatMap(traverserSet);
+                } else {
+                    /*
+                     * Streaming service, mark the next start and call.
```

Comment:
Just to confirm: it looks like batching support (and the resulting requirement of grouping parameters, etc.) is still missing?